service/storage_service.cc

```diff
@@ -6037,6 +6044,67 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
     }
 }
 
+future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+    // The coordinator does not execute global token metadata barrier before jumping to "cleanup" stage, so we need
+    // a barrier here to see the token metadata which is at least as recent as that of the sender.
+    auto& raft_server = _group0->group0_server();
+    co_await raft_server.read_barrier(&_abort_source);
+
+    shard_id shard;
+
+    { // do not keep token metadata longer than necessary in order not to block barriers
+        auto tm = _shared_token_metadata.get();
+        auto& tmap = tm->tablets().get_tablet_map(tablet.table);
+        auto *trinfo = tmap.get_tablet_transition_info(tablet.tablet);
+
+        // Check if the request is still valid.
+        // If there is mismatch, it means this cleanup was canceled and the coordinator moved on.
+        if (!trinfo) {
+            throw std::runtime_error(format("No transition info for tablet {}", tablet));
+        }
+        if (trinfo->stage != locator::tablet_transition_stage::cleanup) {
+            throw std::runtime_error(format("Tablet {} stage is not at cleanup", tablet));
+        }
+
+        auto& tinfo = tmap.get_tablet_info(tablet.tablet);
+        locator::tablet_replica leaving_replica = locator::get_leaving_replica(tinfo, *trinfo);
+        if (leaving_replica.host != tm->get_my_id()) {
+            throw std::runtime_error(format("Tablet {} has leaving replica different than this one", tablet));
+        }
+        auto shard_opt = tmap.get_shard(tablet.tablet, tm->get_my_id());
+        if (!shard_opt) {
+            on_internal_error(slogger, format("Tablet {} has no shard {} on this node", tablet));
+        }
+        shard = *shard_opt;
+    }
+
+    if (_tablet_cleanup[tablet]) {
+        slogger.debug("Cleanup retry joining with existing session for tablet {}", tablet);
+        co_await _tablet_cleanup[tablet]->get_future();
+        co_return;
+    }
+
+    auto async_gate_holder = _async_gate.hold();
+    promise<> p;
+    _tablet_cleanup[tablet] = seastar::shared_future<>(p.get_future());
+    auto erase_tablet_streaming = seastar::defer([&] {
+        _tablet_cleanup.erase(tablet);
+    });
+
+    try {
+        co_await _db.invoke_on(shard, [tablet] (replica::database& db) {
+            auto& table = db.find_column_family(tablet.table);
+            return table.cleanup_tablet(tablet.tablet);
```

Comment:
We might be working on outdated token_metadata, no? We might not be the tablet owner by now.

Reply:
It's not possible that we're not the owner because the coordinator will not move forward with the migration until this completes.